# 中国五个城市PM 2.5数据分析

### 简介

##### PM2.5的检测数值通常包含在环境部门和公司的空气质量报告中。 PM2.5是指大气颗粒物（PM），其直径小于或等于2.5微米。在环境科学中，特指悬浮在空气中的固体颗粒或液滴，是空气污染的主要来源之一。

### 观察数据

本数据包含有中国五个城市(北京、上海、成都、广州、沈阳)从2010/1/1-2015/12/31的空气和气象数据。数据中的缺失值被标记为NaN。数据列名的含义如下：

No: 行号
year: 年份
month: 月份
day: 日期
hour: 小时
season: 季节
PM: PM2.5浓度 (ug/m^3)
DEWP: 露点 (摄氏温度) 指在固定气压之下，空气中所含的气态水达到饱和而凝结成液态水所需要降至的温度。
TEMP: Temperature (摄氏温度)
HUMI: 湿度 (%)
PRES: 气压 (hPa)
cbwd: 组合风向
Iws: 累计风速 (m/s)
precipitation: 降水量/时 (mm)
Iprec: 累计降水量 (mm)

其中与PM2.5浓度相关的数据包含有多列，其中PM_US Post 的数据来源是US Department of State Air Quality Monitoring Program。其他PM相关的数据来源于中国生态环境部在五个城市设置的不同站点，比如PM_Xuhui位于上海市徐汇区，PM_Dongsihuan位于北京市的东四环。

本项目中的数据可以从Kaggle网站上下载到，该数据也被收录于UCI Machine Learning Repository。

由于项目难度设定的问题，在后面的分析中我们暂时没有对气象数据的处理和分析，如果同学感兴趣的话可以自行探索。如果你有足够的能力，我们也欢迎你不采用项目模板中的代码，对数据自行进行分析~

### 数据评估

现在是时候由你自己来探索数据了。我们已经将数据文件提供给你，你无需自己进行下载任何其他文件。此数据由五个文件组成，分别为：

BeijingPM20100101_20151231.csv,
ChengduPM20100101_20151231.csv,
GuangzhouPM20100101_20151231.csv,
ShanghaiPM20100101_20151231.csv,
ShenyangPM20100101_20151231.csv
首先我们先运行下面的代码单元格来加载你将在分析中使用的所有数据包和函数。这些都是在数据分析中常用的第三方库，我们会在后面的正式课程中进行更加细致的教学。

如果处理的数据量特别大时，最好先从数据样本开始。这样更容易检查我们的数据整理步骤是否有效，因为我们完成代码所需的时间将更少。如果我们对整个过程的进展较为满意，那就可以着手整理整个数据集。（注：本数据集的数量并不是特别大，例如50 MB以上，因此我们并未从中提取样本，而是直接对整个数据集进行的分析。）

In [3]:
import csv
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn
%matplotlib inline

以上海数据为例，我们先使用Pandas的read_csv函数导入第一个数据集，并使用head、info、describe方法来查看数据中的基本信息。

In [29]:
Shanghai_data = pd.read_csv('ShanghaiPM20100101_20151231.csv')

In [30]:
Shanghai_data.head(10)

No  year  month  day  hour  season  PM_Jingan  PM_US Post  PM_Xuhui  DEWP  \
0   1  2010      1    1     0       4        NaN         NaN       NaN  -6.0   
1   2  2010      1    1     1       4        NaN         NaN       NaN  -6.0   
2   3  2010      1    1     2       4        NaN         NaN       NaN  -7.0   
3   4  2010      1    1     3       4        NaN         NaN       NaN  -6.0   
4   5  2010      1    1     4       4        NaN         NaN       NaN  -6.0   
5   6  2010      1    1     5       4        NaN         NaN       NaN  -7.0   
6   7  2010      1    1     6       4        NaN         NaN       NaN  -6.0   
7   8  2010      1    1     7       4        NaN         NaN       NaN  -5.0   
8   9  2010      1    1     8       4        NaN         NaN       NaN  -3.0   
9  10  2010      1    1     9       4        NaN         NaN       NaN  -2.0   

    HUMI    PRES  TEMP cbwd   Iws  precipitation  Iprec  
0  59.48  1026.1   1.0   cv   1.0            0.0    0.0  
1  59.48  1025.1   1.0   SE   2.0            0.0    0.0  
2  59.21  1025.1   0.0   SE   4.0            0.0    0.0  
3  63.94  1024.0   0.0   SE   5.0            0.0    0.0  
4  63.94  1023.0   0.0   SE   8.0            0.0    0.0  
5  59.21  1023.0   0.0   SE  11.0            0.0    0.0  
6  59.48  1023.0   1.0   SE  14.0            0.0    0.0  
7  64.18  1023.0   1.0   SE  17.0            0.0    0.0  
8  69.43  1023.0   2.0   SE  20.0            0.0    0.0  
9  64.90  1023.0   4.0   SE  23.0            0.0    0.0

从运行结果可以看出，除了上面提到的数据列之外，上海数据中还包含有PM_Jingan和PM_Xuhui两个观测站点的数据。并且数据中PM2.5的这三列包含有缺失值“NaN”。

接下来我们可以查看一下数据的缺失情况，以及数据的类型：

In [31]:
Shanghai_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52584 entries, 0 to 52583
Data columns (total 17 columns):
No               52584 non-null int64
year             52584 non-null int64
month            52584 non-null int64
day              52584 non-null int64
hour             52584 non-null int64
season           52584 non-null int64
PM_Jingan        24700 non-null float64
PM_US Post       34039 non-null float64
PM_Xuhui         25189 non-null float64
DEWP             52571 non-null float64
HUMI             52571 non-null float64
PRES             52556 non-null float64
TEMP             52571 non-null float64
cbwd             52572 non-null object
Iws              52572 non-null float64
precipitation    48575 non-null float64
Iprec            48575 non-null float64
dtypes: float64(10), int64(6), object(1)
memory usage: 6.8+ MB


##### 通过数据中的打印，我们可以看出本数据中包含有整型(int)、浮点型(float)、字符串型(str)等数据的格式。

In [32]:
# print type of python object
print(type(Shanghai_data['cbwd'][0]))

<class 'str'>


##### 变量名PM_US Post中包含空格，这也可能对我们后续的分析造成一定的困扰。因为大多数命令中，都是默认以空格做为值与值之间的分隔符，而不是做为文件名的一部分。因此我们需要将变量名中的空格改为下划线:

In [34]:
Shanghai_data.columns = [c.replace(' ','_') for c in Shanghai_data.columns]
Shanghai_data.head()

No  year  month  day  hour  season  PM_Jingan  PM_US_Post  PM_Xuhui  DEWP  \
0   1  2010      1    1     0       4        NaN         NaN       NaN  -6.0   
1   2  2010      1    1     1       4        NaN         NaN       NaN  -6.0   
2   3  2010      1    1     2       4        NaN         NaN       NaN  -7.0   
3   4  2010      1    1     3       4        NaN         NaN       NaN  -6.0   
4   5  2010      1    1     4       4        NaN         NaN       NaN  -6.0   

    HUMI    PRES  TEMP cbwd  Iws  precipitation  Iprec  
0  59.48  1026.1   1.0   cv  1.0            0.0    0.0  
1  59.48  1025.1   1.0   SE  2.0            0.0    0.0  
2  59.21  1025.1   0.0   SE  4.0            0.0    0.0  
3  63.94  1024.0   0.0   SE  5.0            0.0    0.0  
4  63.94  1023.0   0.0   SE  8.0            0.0    0.0

##### 其中season的数据类型为int，但实际上season并不是一个数字，而是一个分类变量，我们可以将season中的数值1,2,3,4转换为对应的季节（可以使用Pandas的map方法来实现）：

In [38]:
Shanghai_data['season'] = Shanghai_data['season'].map({1:'Spring', 2:'Summer', 3:'Autumn', 4: 'Winter'})
Shanghai_data.head()

No  year  month  day  hour  season  PM_Jingan  PM_US_Post  PM_Xuhui  DEWP  \
0   1  2010      1    1     0  Winter        NaN         NaN       NaN  -6.0   
1   2  2010      1    1     1  Winter        NaN         NaN       NaN  -6.0   
2   3  2010      1    1     2  Winter        NaN         NaN       NaN  -7.0   
3   4  2010      1    1     3  Winter        NaN         NaN       NaN  -6.0   
4   5  2010      1    1     4  Winter        NaN         NaN       NaN  -6.0   

    HUMI    PRES  TEMP cbwd  Iws  precipitation  Iprec  
0  59.48  1026.1   1.0   cv  1.0            0.0    0.0  
1  59.48  1025.1   1.0   SE  2.0            0.0    0.0  
2  59.21  1025.1   0.0   SE  4.0            0.0    0.0  
3  63.94  1024.0   0.0   SE  5.0            0.0    0.0  
4  63.94  1023.0   0.0   SE  8.0            0.0    0.0

#####  并且通过打印数据的总行数，与上方non-null的数据数量进行对比，我们可以发现一些数据中存在缺失值，以及缺失值的数量。比如PM_Jingan中的non-null非空数据行数为24700, 该数据集的总行数为52584条，该数据缺失的条数则为：27884。

说明：下面代码中用到的dropna()函数能够删除掉数据中的缺失值。

In [40]:
# print the length of data
print("The number of row in this dataset is ", len(Shanghai_data.index))

# calculating the number of records in column "PM_Jingan"
print("There number of missing data records in PM_Jingan is: ",
     len(Shanghai_data.index) - len(Shanghai_data['PM_Jingan'].dropna()))

The number of row in this dataset is  52584
There number of missing data records in PM_Jingan is:  27884


#####  请选择一个你感兴趣的城市数据集，对你的数据进行类似于上方的评估，并回答下面的问题，说明你观察到的信息：

In [44]:
Beijing_data = pd.read_csv('BeijingPM20100101_20151231.csv')
Beijing_data.head(10)

No  year  month  day  hour  season  PM_Dongsi  PM_Dongsihuan  \
0   1  2010      1    1     0       4        NaN            NaN   
1   2  2010      1    1     1       4        NaN            NaN   
2   3  2010      1    1     2       4        NaN            NaN   
3   4  2010      1    1     3       4        NaN            NaN   
4   5  2010      1    1     4       4        NaN            NaN   
5   6  2010      1    1     5       4        NaN            NaN   
6   7  2010      1    1     6       4        NaN            NaN   
7   8  2010      1    1     7       4        NaN            NaN   
8   9  2010      1    1     8       4        NaN            NaN   
9  10  2010      1    1     9       4        NaN            NaN   

   PM_Nongzhanguan  PM_US Post  DEWP  HUMI    PRES  TEMP cbwd    Iws  \
0              NaN         NaN -21.0  43.0  1021.0 -11.0   NW   1.79   
1              NaN         NaN -21.0  47.0  1020.0 -12.0   NW   4.92   
2              NaN         NaN -21.0  43.0  1019.0 -11.0   NW   6.71   
3              NaN         NaN -21.0  55.0  1019.0 -14.0   NW   9.84   
4              NaN         NaN -20.0  51.0  1018.0 -12.0   NW  12.97   
5              NaN         NaN -19.0  47.0  1017.0 -10.0   NW  16.10   
6              NaN         NaN -19.0  44.0  1017.0  -9.0   NW  19.23   
7              NaN         NaN -19.0  44.0  1017.0  -9.0   NW  21.02   
8              NaN         NaN -19.0  44.0  1017.0  -9.0   NW  24.15   
9              NaN         NaN -20.0  37.0  1017.0  -8.0   NW  27.28   

   precipitation  Iprec  
0            0.0    0.0  
1            0.0    0.0  
2            0.0    0.0  
3            0.0    0.0  
4            0.0    0.0  
5            0.0    0.0  
6            0.0    0.0  
7            0.0    0.0  
8            0.0    0.0  
9            0.0    0.0

In [45]:
Beijing_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52584 entries, 0 to 52583
Data columns (total 18 columns):
No                 52584 non-null int64
year               52584 non-null int64
month              52584 non-null int64
day                52584 non-null int64
hour               52584 non-null int64
season             52584 non-null int64
PM_Dongsi          25052 non-null float64
PM_Dongsihuan      20508 non-null float64
PM_Nongzhanguan    24931 non-null float64
PM_US Post         50387 non-null float64
DEWP               52579 non-null float64
HUMI               52245 non-null float64
PRES               52245 non-null float64
TEMP               52579 non-null float64
cbwd               52579 non-null object
Iws                52579 non-null float64
precipitation      52100 non-null float64
Iprec              52100 non-null float64
dtypes: float64(11), int64(6), object(1)
memory usage: 7.2+ MB


#### Beijing_data.info()

#####  问题2a: 你选择的城市数据中包含哪几个PM 2.5的监测点。请说明它们的列名，不需要包含PM_US Post.
#####  答案：PM_Dongsi PM_Dongsihuan PM_Nongzhanguan
#####  问题2b: 你选择的城市数据中的哪些列存在缺失值？缺失值的数量分别是多少？


#####  问题2b: 你选择的城市数据中的哪些列存在缺失值？缺失值的数量分别是多少？

In [53]:
# print the length of data 
print("The number of row in this dataset is ", len(Beijing_data.index))
# calculating the number of records in column "PM_Dongsi"
print("There number of missing data records in PM_Dongsi is: ",
     len(Beijing_data.index) - len(Beijing_data['PM_Dongsi'].dropna()))
print("-----------------------------------------------------------------")
# print the length of data 
print("The number of row in this dataset is ", len(Beijing_data.index))
# calculating the number of records in column "PM_Dongsihuan"
print("There number of missing data records in PM_Dongsihuani is: ",
     len(Beijing_data.index) - len(Beijing_data['PM_Dongsihuan'].dropna()))
print("-----------------------------------------------------------------")
# print the length of data 
print("The number of row in this dataset is ", len(Beijing_data.index))
# calculating the number of records in column "PM_Nongzhanguan"
print("There number of missing data records in PM_Nongzhanguan is: ",
     len(Beijing_data.index) - len(Beijing_data['PM_Nongzhanguan'].dropna()))

The number of row in this dataset is  52584
There number of missing data records in PM_Dongsi is:  27532
-----------------------------------------------------------------
The number of row in this dataset is  52584
There number of missing data records in PM_Dongsihuani is:  32076
-----------------------------------------------------------------
The number of row in this dataset is  52584
There number of missing data records in PM_Nongzhanguan is:  27653


### 数据整理

##### 现在你已使用单个数据集完成了一些探索，是时候更进一步，将所有数据整理到一个文件中并看看你能发现什么趋势。通过describe函数对数据进行查看，我们可以看出几个PM 2.5观察站的统计数据都很接近，经过进一步的分析，我们会能够发现这几个观测站的数据存在有很强的相关关系（本项目中并未包含，但你可以通过学习后面的统计学课程，自己来完成此部分）。

因为五个数据文件中都包含PM_US Post一列，并且该列的缺失值相对于其他列缺失值较小，因此在下面的分析中我们仅保留该列数据作为PM 2.5的关键数据。在下面的代码中我们也预先对所有城市的season进行了转换，并为数据添加了一个city列，便于对不同城市进行对比分析。

其他的气象数据我们暂时先不考虑，如果同学感兴趣的话可以自行分析。